### PR TITLE
[Frontend] Enzyme integration

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** Enabling differentiation of callbacks in the frontend.

**Description of the Change:** Adds:

1. `CallbackWithPotentialCustomGrad` class which all `pure_callback`s are.
2. `CallbackWithCustomGrad` class which only `pure_callback`s with custom grads are.
3. Create MLIR operations during lowering in the frontend.
4. Cache MLIR operations appropriately.
5. Modify analysis that verifies grad op to allow `CallbackCallOp`s which:
    * Do not return anything (inactive callbacks)
    * Use a `CallbackOp` symbol which is used in `CustomGradOp`.

**Benefits:** Enabling differentiation for callbacks.
[sc-60499]

### CODE REVIEW

File: frontend/catalyst/api_extensions/callbacks.py [hunk 1]
```diff
@@ -39,6 +40,95 @@
...
+        # Where does the infinite recursion happen?
+        # It happens if the fwd or bwd passes have a call to
+        # the pure_callback implementation.
+        self.callback = base_callback(closure, custom_grad=self)
+
+        # The arguments here are tracers.
+        # And we want to just get the abstraction of the tracers (i.e., the types)
+        absargs, abskwargs = tree_map(shaped_abstractify, (args, kwargs))
+        # Once we have the types, we can call this self.func with the absargs and abskwargs.
+        # We don't need the jaxpr representation but the output is the shape of the cotangents.
```
Comment (erick-xanadu) [inline frontend/catalyst/api_extensions/callbacks.py:73]: Use self.callback instead of self.

File: frontend/catalyst/api_extensions/callbacks.py [hunk 2]
```diff
@@ -39,6 +40,95 @@
...
+        _jaxpr, cotangents = jax.make_jaxpr(self, return_shape=True)(*absargs, **abskwargs)
```
Comment (erick-xanadu) [inline frontend/catalyst/api_extensions/callbacks.py:74]: Are the cotangents not just going to be `self.restype` always?

File: mlir/lib/Quantum/Transforms/annotate_function.cpp [hunk 3]
```diff
@@ -35,14 +37,34 @@ bool isAnnotated(FunctionOpInterface op, const char *attr)
...
+                SymbolTable::lookupNearestSymbolFrom<FunctionOpInterface>(callbackCall, callee);
+            bool inactive = callback.getResultTypes().empty();
+            if (inactive) {
+                return WalkResult::advance();
+            }
+            auto uses = *SymbolTable::getSymbolUses(callback, mod);
+            for (SymbolTable::SymbolUse use : uses) {
+                Operation *user = use.getUser();
+                if (auto customGrad = dyn_cast<catalyst::gradient::CustomGradOp>(user)) {
+                    isDifferentiated |= customGrad.getCalleeAttr() == callee;
```
Comment (dime10) [inline mlir/lib/Quantum/Transforms/annotate_function.cpp:58]: Why is the callee compared, don't we know that this CustomGradOp is already a user of the callback?

File: frontend/test/pytest/test_callback.py
Comment (dime10) [inline frontend/test/pytest/test_callback.py:1]: Thanks for the tests! My only concern is that the identity function could hide bugs like incorrect values (only appearing in non-trivial gradients), or swapped dimensions (like producing (2,3) instead of (3,2)). 

Typically I would try to go for asymmetric input and outputs and ensure values (including results) are different from 0 and 1.

**Review by dime10 — CHANGES_REQUESTED** [review]
Nice work!

File: frontend/test/pytest/test_callback.py [hunk 4]
```diff
@@ -586,5 +587,269 @@ def wrapper():
...
+@pytest.mark.parametrize("arg0", [(0.1), (0.2), (0.3)])
+@pytest.mark.parametrize("arg1", [(2.0), (3.0), (4.0)])
+def test_example_from_story(arg0, arg1):
+    """Just exactly the same function on the spec
+    modulo errors in the example
+    """
+
+    @pure_callback
+    def some_func(x, y) -> float:
+        return np.sin(x) * y
```
Comment (josh146) [inline frontend/test/pytest/test_callback.py:753]: Hey @erick-xanadu, apologies for not catching in the in the story, but it would be great to copy and paste this test with a slight modification, where the input is an array:

```python
@pure_callback
def some_func(x) -> float:
    return np.sin(x[0]) * x[1]

@some_func.fwd
def some_func_fwd(x):
    return some_func(x), (jnp.cos(x[0]), jnp.sin(x[0]), x[1])

@some_func.bwd
def some_func_bws(res, dy):
    cos_x0, sin_x0, y0 = res  # Gets residuals computed in f_fwd
    # since there is a single array parameter, we return
    # a VJP which is an array parameter of the same shape
    return jnp.array([cos_x * dy * y, sin_x * dy])

@qml.qjit
@grad
def cost(x):
    y = jnp.array([jnp.cos(x[0]), x[1]])
    return jnp.sin(some_func(y))
```

Let me know if by some chance this isn't working (no need to dig into it), and I'll track it on SC.

Comment (josh146) [reply frontend/test/pytest/test_callback.py:753]: Another one that could be good to add a test for is `some_func(x: Array, y: Array) -> Array`, but I'll need to craft an example of this so no worries on this one.

Comment (josh146) [reply frontend/test/pytest/test_callback.py:753]: @erick-xanadu I've written up all test cases I can think of here: https://gist.github.com/josh146/b65436b71221a12df9a1ad9413df2314

File: frontend/test/pytest/test_callback.py
Comment (erick-xanadu) [reply frontend/test/pytest/test_callback.py:1]: I'll add these extra tests.

Comment (erick-xanadu) [reply frontend/test/pytest/test_callback.py:1]: Can you check to see if with Josh's extra tests this satisfies your criteria?

File: frontend/test/pytest/test_callback.py (hunk 4 above)
Comment (erick-xanadu) [reply frontend/test/pytest/test_callback.py:754]: I've added them all with minor differences. None of the differences seem to be significant (i.e., the error appears to be in the test not on the code). But it might be worthwhile to double check.

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 3 above)
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp:58]: `CustomGradOp` has three different symbols. The callee, the forward and reverse. Just because `CustomGradOp` uses the same symbol the callback uses, it doesn't mean that it is guaranteed to be in the callee.

File: frontend/test/pytest/test_callback.py (hunk 4 above)
Comment (erick-xanadu) [reply frontend/test/pytest/test_callback.py:754]: I'll also add an issue to add a warning if the user uses a shape with dtype=float32. I am pretty sure we must only use float64.

File: frontend/test/pytest/test_callback.py [hunk 5]
```diff
@@ -811,6 +812,330 @@ def wrapper_jit(x):
...
+
+    @some_func.bwd
+    def some_func_bwd(res, dy):
+        x, y = res
+        vjp0 = y * jnp.cos(x) * jnp.reshape(dy[0], (-1, 1))
+        vjp1 = dy[0] @ jnp.sin(x) + 2 * y * dy[1]
+        return (vjp0, vjp1)
+
+    @qml.qjit
+    @partial(grad, argnum=[0, 1])  # We just use argnum instead of argnums?
```
Comment (josh146) [inline frontend/test/pytest/test_callback.py:1019]: oh, do we? Huh.

File: frontend/catalyst/api_extensions/callbacks.py [hunk 6]
```diff
@@ -39,6 +40,95 @@
...
+        self._bwd = reverse
+        self._bwd_jaxpr = None
+        self.callback = None
+
+    def __call__(self, *args, **kwargs):
+        if self.callback:
+            return self.callback(*args, **kwargs)
+
+        def closure(*args, **kwargs) -> self.restype:
+            return self.func(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py:61]: No, it is not necessarily annotated. We use an alternative functional syntax in that case.

File: frontend/test/pytest/test_callback.py [hunk 7]
```diff
@@ -811,6 +812,330 @@ def wrapper_jit(x):
...
+
+    @some_func.fwd
+    def some_func_fwd(weights):
+        return some_func(weights), weights
+
+    @pure_callback
+    def some_func_bwd_vjp(res, dy) -> vjp_shape:
+        vjp0 = res["y"] * jnp.cos(res["x"]) * jnp.reshape(dy["one"], (-1, 1))
+        vjp1 = dy["one"] @ jnp.sin(res["x"]) + 2 * res["y"] * dy["two"]
+        return ({"x": vjp0, "y": vjp1},)
```
Comment (josh146) [inline frontend/test/pytest/test_callback.py:1114]: @erick-xanadu since this is a pure callback, I thought it would be important to use `np.` here (since we know numpy won't work without the purecallback)

File: frontend/catalyst/api_extensions/callbacks.py (hunk 1 above)
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py]: https://github.com/PennyLaneAI/catalyst/pull/834/commits/c7b80d3fce9f7b5c8a08b1325a4fdcdac2407120

File: frontend/catalyst/api_extensions/callbacks.py (hunk 2 above)
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py]: https://github.com/PennyLaneAI/catalyst/pull/834/commits/c7b80d3fce9f7b5c8a08b1325a4fdcdac2407120

File: frontend/catalyst/api_extensions/callbacks.py [hunk 8]
```diff
@@ -39,6 +40,95 @@
 from catalyst.utils.types import convert_pytype_to_shaped_array
 
 
+class CallbackWithCustomGrad:
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py]: https://github.com/PennyLaneAI/catalyst/pull/834/commits/263ad9af87b167bbefac1b328eabf0aed422978e

File: mlir/lib/Quantum/Transforms/annotate_function.cpp [hunk 9]
```diff
@@ -35,14 +37,34 @@ bool isAnnotated(FunctionOpInterface op, const char *attr)
...
-            dyn_cast<catalyst::CallbackCallOp>(o)) {
+    ModuleOp mod = op->getParentOfType<ModuleOp>();
+    auto res = op.walk([&](Operation *o) {
+        if (dyn_cast<MeasureOp>(o) || dyn_cast<catalyst::CustomCallOp>(o)) {
             return WalkResult::interrupt();
         }
-        else {
-            return WalkResult::advance();
+        else if (auto callbackCall = dyn_cast<catalyst::CallbackCallOp>(o)) {
+            bool isDifferentiated = false;
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: https://github.com/PennyLaneAI/catalyst/pull/834/commits/2c8ee0fd237737e4febbf8a0dd2cd85bd6d2e044

File: frontend/catalyst/api_extensions/callbacks.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/callbacks.py]: Can we replace `func` at the point where we detect the functional syntax then? That way we don't need to replicate this wrapper in a few places and the logic for it is more localized. 
I'm also finding the name somewhat confusing because we don't appear to close over any variables besides the type annotation.

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 3 above)
Comment (dime10) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: Makes sense, thanks!

File: frontend/test/pytest/test_callback.py
Comment (dime10) [reply frontend/test/pytest/test_callback.py:1]: Yes those tests are great! The only missing I think is to test array in/out but with different dimensions (e.g. 2 elements in, 3 out). And also to assert the shape as well as the values.

File: frontend/test/pytest/test_callback.py (hunk 5 above)
Comment (dime10) [reply frontend/test/pytest/test_callback.py:1021]: Yes unfortunate convention we took from pennylane, we should definitely update this to be `argnums`.

File: frontend/test/pytest/test_callback.py
Comment (josh146) [reply frontend/test/pytest/test_callback.py:1]: Good point!

Comment (erick-xanadu) [reply frontend/test/pytest/test_callback.py:1]: Asserting shapes match: https://github.com/PennyLaneAI/catalyst/pull/834/commits/520e0649b2d340e6ee367a014a11b5f58e965905

File: frontend/catalyst/api_extensions/callbacks.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/callbacks.py]: Can we get this in and then I can clean this bit up? I just don't want to add any extra time. This seems like a really small issue that (while I agree can be solved here) I just want to get this feature in. I also need time to work on QSVT.

File: frontend/test/pytest/test_callback.py (hunk 4 above)
Comment (erick-xanadu) [reply frontend/test/pytest/test_callback.py:756]: I added the one requested by David, and I tried also differentiating A @ B and comparing it to JAX. The second one is currently producing wrong results. This could be related to the same issue we've been seeing in the past regarding not passing metadata to Enzyme but I am not 100% sure yet.

Comment (erick-xanadu) [reply frontend/test/pytest/test_callback.py:756]: https://github.com/PennyLaneAI/catalyst/pull/834/commits/c1916125a8972c374de161f554089e394996789c

File: frontend/test/pytest/test_callback.py [hunk 10]
```diff
@@ -586,5 +590,696 @@ def wrapper():
...
+
+    def fun(x):
+        y = jnp.array([3.0, 4.0, 5.0])
+        return y * x[0] + x[1]
+
+    # hack for the type
+    ty = jnp.array([2.0, 2.0, 2.0])
+
+    @pure_callback
+    def fun_callback(x) -> ty:
```
Comment (dime10) [inline frontend/test/pytest/test_callback.py:1155]: @josh146 this makes me think we really need a simpler of specifying array types 😅

Comment (erick-xanadu) [reply frontend/test/pytest/test_callback.py:1155]: Jax suffers from the same issue :thinking: Here, though, I always know that a value can be mapped to a type, so I just make values because I don't like making ShapedArray or the other version that Josh uses :sweat_smile:

File: frontend/catalyst/api_extensions/callbacks.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/callbacks.py]: Tbh the smaller the issue the less likely it is going to be picked in the future. I already forgot the issues that were left for later even in just this series of PRs and I doubt we'll ever come back to those either.

But I won't reject the PR over it.

File: frontend/test/pytest/test_callback.py (hunk 10 above)
Comment (dime10) [reply frontend/test/pytest/test_callback.py:1155]: Absolutely, that doesn't mean we couldn't improve it though :)

Comment (dime10) [reply frontend/test/pytest/test_callback.py:1155]: For comparison, the ShapedArray way `jax.core.ShapedArray((3,), float)`. It does feel annoying to type, I agree, although I'm not sure why, it's only slightly longer in this case. 

And it does seem harder to do better than `fun(shape, dtype)` 🤔

**Review by dime10 — APPROVED** [review]
🥳

File: frontend/test/pytest/test_callback.py (hunk 4 above)
Comment (josh146) [reply frontend/test/pytest/test_callback.py:756]: Thanks @erick-xanadu! Could you open an issue for the failing test example if you haven't already (too many notifications atm 😅)
